### PR TITLE
More const everywhere

### PR DESCRIPTION
**Pull request by ThomasHabets**
Well… the parts I dove into so far.

### CODE REVIEW

**Comment by mbr0wn** [comment]
Who doesn't love const?

File: gr-digital/lib/costas_loop_cc_impl.cc [hunk 1]
```diff
@@ -50,43 +50,36 @@ costas_loop_cc_impl::costas_loop_cc_impl(float loop_bw, int order, bool use_snr)
...
                     boost::bind(&costas_loop_cc_impl::handle_set_noise, this, _1));
 }
 
 costas_loop_cc_impl::~costas_loop_cc_impl() {}
 
+costas_loop_cc_impl::d_phase_detector_t costas_loop_cc_impl::choose_phase_detector(int order, bool use_snr)
+{
+  switch (order) {
+  case 2:
+    if (use_snr)
```
Comment (mbr0wn) [inline gr-digital/lib/costas_loop_cc_impl.cc]: If you're already touching this, please add {} so clang-tidy won't freak out once we enable that.

Comment (ThomasHabets) [reply gr-digital/lib/costas_loop_cc_impl.cc]: Oooh, braces around every `if`? Great!

File: gr-digital/lib/costas_loop_cc_impl.cc [hunk 2]
```diff
@@ -103,7 +99,7 @@ float costas_loop_cc_impl::phase_detector_8(gr_complex sample) const
        Circuits and Systems, Vol. 2, pp. 1447 - 1450, 2004.
     */
 
-    float K = (sqrt(2.0) - 1);
+    const float K = (sqrt(2.0) - 1);
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.cc]: should be `sqrtf`, not `sqrt`. The back and forth between double and float has no upside, and sadly can't be optimized away.

File: gr-digital/lib/costas_loop_cc_impl.cc [hunk 3]
```diff
@@ -126,8 +122,8 @@ float costas_loop_cc_impl::phase_detector_2(gr_complex sample) const
 
 float costas_loop_cc_impl::phase_detector_snr_8(gr_complex sample) const
 {
-    float K = (sqrt(2.0) - 1);
-    float snr = abs(sample) * abs(sample) / d_noise;
+    const float K = (sqrt(2.0) - 1);
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.cc]: same as above.

File: gr-digital/lib/costas_loop_cc_impl.cc (hunk 2 above)
Comment (marcusmueller) [reply gr-digital/lib/costas_loop_cc_impl.cc]: could you fix that in a separate commit, please?

File: gr-digital/lib/costas_loop_cc_impl.cc [hunk 4]
```diff
@@ -126,8 +122,8 @@ float costas_loop_cc_impl::phase_detector_2(gr_complex sample) const
...
+    const float snr = abs(sample) * abs(sample) / d_noise;
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.cc]: |z|·|z| = z·conj(z); please replace by 

`const float snr = sample * std::conj(sample) / d_noise;` 

for superior compute.

File: gr-digital/lib/costas_loop_cc_impl.cc [hunk 5]
```diff
@@ -139,14 +135,14 @@ float costas_loop_cc_impl::phase_detector_snr_8(gr_complex sample) const
 
 float costas_loop_cc_impl::phase_detector_snr_4(gr_complex sample) const
 {
-    float snr = abs(sample) * abs(sample) / d_noise;
+    const float snr = abs(sample) * abs(sample) / d_noise;
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.cc]: see above

File: gr-digital/lib/costas_loop_cc_impl.cc [hunk 6]
```diff
@@ -139,14 +135,14 @@ float costas_loop_cc_impl::phase_detector_snr_8(gr_complex sample) const
...
     return ((blocks::tanhf_lut(snr * sample.real()) * sample.imag()) -
             (blocks::tanhf_lut(snr * sample.imag()) * sample.real()));
 }
 
 float costas_loop_cc_impl::phase_detector_snr_2(gr_complex sample) const
 {
-    float snr = abs(sample) * abs(sample) / d_noise;
+    const float snr = abs(sample) * abs(sample) / d_noise;
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.cc]: see above

File: gr-digital/lib/costas_loop_cc_impl.h [hunk 7]
```diff
@@ -32,7 +32,7 @@ namespace digital {
 class costas_loop_cc_impl : public costas_loop_cc
 {
 private:
-    int d_order;
+    const int d_order;
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.h]: make it unsigned.

File: gr-digital/lib/costas_loop_cc_impl.h [hunk 8]
```diff
@@ -85,8 +85,9 @@ class costas_loop_cc_impl : public costas_loop_cc
      */
     float phase_detector_snr_2(gr_complex sample) const; // for BPSK
 
-
-    float (costas_loop_cc_impl::*d_phase_detector)(gr_complex sample) const;
+    typedef float (costas_loop_cc_impl::*d_phase_detector_t)(gr_complex sample) const;
+    static d_phase_detector_t choose_phase_detector(int order, bool use_snr);
```
Comment (marcusmueller) [inline gr-digital/lib/costas_loop_cc_impl.h]: unsigned

File: gr-digital/lib/diff_encoder_bb_impl.cc [hunk 9]
```diff
@@ -54,10 +54,9 @@ int diff_encoder_bb_impl::work(int noutput_items,
     unsigned char* out = (unsigned char*)output_items[0];
 
     unsigned last_out = d_last_out;
-    unsigned modulus = d_modulus;
 
     for (int i = 0; i < noutput_items; i++) {
```
Comment (marcusmueller) [inline gr-digital/lib/diff_encoder_bb_impl.cc]: while you're at it: positive loop variable can happily be `unsigned int i`.

File: gr-digital/lib/fll_band_edge_cc_impl.cc [hunk 10]
```diff
@@ -144,7 +144,7 @@ void fll_band_edge_cc_impl::design_filter(float samps_per_sym,
                                           float rolloff,
                                           int filter_size)
 {
-    int M = rint(filter_size / samps_per_sym);
+    const int M = rint(filter_size / samps_per_sym);
```
Comment (marcusmueller) [inline gr-digital/lib/fll_band_edge_cc_impl.cc]: should be `rintf`

File: gr-blocks/lib/file_descriptor_sink_impl.cc [hunk 11]
```diff
@@ -61,7 +61,7 @@ int file_descriptor_sink_impl::work(int noutput_items,
                                     gr_vector_const_void_star& input_items,
                                     gr_vector_void_star& output_items)
 {
-    char* inbuf = (char*)input_items[0];
+    const char* inbuf = (char*)input_items[0];
```
[the remark this answers is not in the document]
Comment (ThomasHabets) [reply gr-blocks/lib/file_descriptor_sink_impl.cc]: Oopsie. Switching to C++ style cast to better catch it in the future.

File: gr-blocks/lib/unpacked_to_packed_impl.cc [hunk 12]
```diff
@@ -79,9 +79,9 @@ unsigned int unpacked_to_packed_impl<T>::get_bit_be1(const T* in_vector,
                                                      unsigned int bit_addr,
                                                      unsigned int bits_per_chunk)
 {
-    unsigned int byte_addr = (int)bit_addr / bits_per_chunk;
-    T x = in_vector[byte_addr];
-    unsigned int residue = bit_addr - byte_addr * bits_per_chunk;
+    const unsigned int byte_addr = (int)bit_addr / bits_per_chunk;
```
[the remark this answers is not in the document]
Comment (ThomasHabets) [reply gr-blocks/lib/unpacked_to_packed_impl.cc]: Yeah that is silly. Removed.

File: gr-blocks/include/gnuradio/blocks/log2_const.h [hunk 13]
```diff
@@ -34,64 +34,64 @@ namespace gr {
...
-static inline int log2_const()
+static inline constexpr int log2_const()
 {
     assert(0);
     return 0;
 }
 
 template <>
-inline int log2_const<1>()
+inline constexpr int log2_const<1>()
```
[the remark this answers is not in the document]
Comment (ThomasHabets) [reply gr-blocks/include/gnuradio/blocks/log2_const.h]: Indeed. I went for "one thing at a time" in this PR though. :-)

**Comment by ThomasHabets** [comment]
Should be all comments so far address.

File: gr-blocks/include/gnuradio/blocks/log2_const.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (mbr0wn) [reply gr-blocks/include/gnuradio/blocks/log2_const.h]: I think there's no action for this PR.

**Comment by mbr0wn** [comment]
@marcusmueller Looks like all your comments are fixed.

**Pull request merged**